sw/device/lib/testing/dif_uart_test.cc

```diff
@@ -0,0 +1,819 @@
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, IrqsRestoreSuccess) {
+  uint32_t state = 0b11111111'11111111'11111111'11111111;
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, state);
+
+  EXPECT_TRUE(dif_uart_irqs_restore(&dif_uart_, state));
+
+  state = 0b00000000'00000000'00000000'00000000;
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, state);
+
+  EXPECT_TRUE(dif_uart_irqs_restore(&dif_uart_, state));
+}
+
+// dif_uart_irq_enable() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInterrupt, IrqEnableFailureUartNull) {
+  EXPECT_FALSE(dif_uart_irq_enable(nullptr, kDifUartInterruptTxWatermark,
+                                   kDifUartEnable));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInterrupt, IrqEnableSuccess) {
+  // Enable interrupts
+  for (auto const& irq : uart_irq_reg_map_) {
+    EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET,
+                  0b00000000'00000000'00000000'00000000);
+    EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, irq.second);
+
+    EXPECT_TRUE(dif_uart_irq_enable(&dif_uart_, irq.first, kDifUartEnable));
+  }
+
+  // Disable interrupts
+  for (auto const& irq : uart_irq_reg_map_) {
+    EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET,
+                  0b11111111'11111111'11111111'11111111);
+    EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, ~irq.second);
+
+    EXPECT_TRUE(dif_uart_irq_enable(&dif_uart_, irq.first, kDifUartDisable));
+  }
+
+  EXPECT_EQ(uart_irq_reg_map_.size(), 8); // TODO
+}
+
+// dif_uart_irq_force() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInterrupt, IrqForceFailureUartNull) {
+  EXPECT_FALSE(dif_uart_irq_force(nullptr, kDifUartInterruptTxWatermark));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInterrupt, IrqForceSuccess) {
+  for (auto const& irq : uart_irq_reg_map_) {
+    EXPECT_READ32(UART_INTR_TEST_REG_OFFSET,
+                  0b00000000'00000000'00000000'00000000);
+    EXPECT_WRITE32(UART_INTR_TEST_REG_OFFSET, irq.second);
+
+    EXPECT_TRUE(dif_uart_irq_force(&dif_uart_, irq.first));
+  }
+
+  EXPECT_EQ(uart_irq_reg_map_.size(), 8); // TODO
+}
+
+// dif_uart_rx_bytes_available() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, RxBytesAvailableFailureUartNull) {
+  size_t num_bytes;
+  EXPECT_FALSE(dif_uart_rx_bytes_available(nullptr, &num_bytes));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, RxBytesAvailableFailureNumBytesNull) {
+  EXPECT_FALSE(dif_uart_rx_bytes_available(&dif_uart_, nullptr));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, RxBytesAvailableSuccess) {
+  // 32 bytes available to read
+  EXPECT_READ32(UART_FIFO_STATUS_REG_OFFSET,
+                 0b01010101'00100000'01010101'01010101);
+
+  size_t num_bytes = 0;
+  EXPECT_TRUE(dif_uart_rx_bytes_available(&dif_uart_, &num_bytes));
+
+  EXPECT_EQ(num_bytes, 32);
+
+  // 0 bytes available to read
+  EXPECT_READ32(UART_FIFO_STATUS_REG_OFFSET,
+                 0b01010101'00000000'01010101'01010101);
+
+  num_bytes = 0xffffffff;
```

Comment:
I'd avoid reusing out-parameter variables.

Reply:
Done. Split up in two tests.